src/enum.lisp

```diff
@@ -152,99 +241,94 @@
-       (notice-foreign-type
-        ',name (make-foreign-bitfield ',name ',base-type ',masks)))))
+  (%defcenum-like name-and-options masks 'make-foreign-bitfield))
 
 (defun foreign-bitfield-symbol-list (bitfield-type)
   "Return a list of SYMBOLS defined in BITFIELD-TYPE."
-  (hash-keys-to-list (symbol-values (parse-type bitfield-type))))
+  (field-keywords (ensure-parsed-base-type bitfield-type)))
 
 (defun %foreign-bitfield-value (type symbols)
-  (reduce #'logior symbols
-          :key (lambda (symbol)
-                 (check-type symbol symbol)
-                 (or (gethash symbol (symbol-values type))
-                     (error "~S is not a valid symbol for bitfield type ~S."
-                            symbol type)))))
+  (declare (optimize speed))
+  (labels ((process-one (symbol)
+             (check-type symbol symbol)
+             (or (gethash symbol (keyword-values type))
+                 (error "~S is not a valid symbol for bitfield type ~S."
+                        symbol type))))
+    (declare (dynamic-extent #'process-one))
+    (if (consp symbols)
+        (reduce #'logior symbols :key #'process-one)
+        (process-one symbols))))
+
+(defmacro %foreign-bitfield-value/form (type symbols)
+  `(let ((type-obj (ensure-parsed-base-type ,type)))
+     (assert (typep type-obj 'foreign-bitfield) ()
+             "~S is not a foreign bitfield type." ,type)
+     (%foreign-bitfield-value type-obj ,symbols)))
 
 (defun foreign-bitfield-value (type symbols)
   "Convert a list of symbols into an integer according to the TYPE bitfield."
-  (let ((type-obj (parse-type type)))
-    (if (not (typep type-obj 'foreign-bitfield))
-      (error "~S is not a foreign bitfield type." type)
-      (%foreign-bitfield-value type-obj symbols))))
+  (%foreign-bitfield-value/form type symbols))
 
 (define-compiler-macro foreign-bitfield-value (&whole form type symbols)
   "Optimize for when TYPE and SYMBOLS are constant."
   (if (and (constantp type) (constantp symbols))
-      (let ((type-obj (parse-type (eval type))))
-        (if (not (typep type-obj 'foreign-bitfield))
-            (error "~S is not a foreign bitfield type." type)
-            (%foreign-bitfield-value type-obj (eval symbols))))
+      (%foreign-bitfield-value/form (eval type) (eval symbols))
       form))
 
 (defun %foreign-bitfield-symbols (type value)
   (check-type value integer)
-  (loop for mask being the hash-keys in (value-symbols type)
-            using (hash-value symbols)
-        when (= (logand value mask) mask)
-        append symbols))
+  (check-type type foreign-bitfield)
+  (loop
+    :with bit-index->keyword = (bit-index->keyword type)
+    :for bit-index :from 0 :below (array-dimension bit-index->keyword 0)
+    :for mask = 1 :then (ash mask 1)
+    :for key = (aref bit-index->keyword bit-index)
+    :when (and key
+               (= (logand value mask) mask))
+    :collect key))
+
+(defmacro %foreign-bitfield-symbols/form (type value)
```

Comment:
This probably shouldn't be a macro either. And it's capturing the type variable by chance in the error form, right?